src/mlpack/tests/pso_test.cpp

```diff
@@ -0,0 +1,246 @@
+#include <mlpack/core/optimizers/problems/bukin_function.cpp>
+#include <mlpack/core/optimizers/problems/matyas_function.cpp>
+#include <mlpack/core/optimizers/problems/colville_function.cpp>
+#include <mlpack/core/optimizers/problems/sgd_test_function.hpp>
+#include <mlpack/core/optimizers/problems/eggholder_function.cpp>
+#include <mlpack/core/optimizers/problems/rosenbrock_function.cpp>
+#include <mlpack/core/optimizers/problems/mc_cormick_function.cpp>
+
+#include <mlpack/methods/logistic_regression/logistic_regression.hpp>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace mlpack;
+using namespace mlpack::optimization;
+using namespace mlpack::distribution;
+using namespace mlpack::regression;
+
+BOOST_AUTO_TEST_SUITE(PSOTest);
+
+/**
+ * Simple test case for PSO.
+ */
+BOOST_AUTO_TEST_CASE(SimpleTest)
+{
+  PSO optimizer;
+  PSOTestFunction f;
+
+  arma::mat iterate;
+  iterate << 0.5828 << 0.0654 << 0.8817;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_SMALL(result, 1e-5);
+}
+
+/**
+ * Test for minimizing the Rosenbrock function.
+ */
+BOOST_AUTO_TEST_CASE(RosenbrockFunctionTest)
+{
+  PSO optimizer;
+  RosenbrockFunction f;
+
+  arma::mat iterate;
+  iterate << 0.5828 << 0.0654 << 0.8817 << 0.3461 << arma::endr
+          << 0.2351 << 0.0034 << 0.1641 << 0.7621;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_SMALL(result, 1e-5);
+}
+
+/**
+ * Test for minimizing Matyas function.
+ */
+BOOST_AUTO_TEST_CASE(MatyasFunctionTest)
+{
+  PSO optimizer;
+  MatyasFunction f;
+
+  arma::mat iterate;
+  iterate << 0.5828 << 0.0654 << 0.8817 << 0.3461 << arma::endr
+          << 0.2351 << 0.0034 << 0.1641 << 0.7621;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_SMALL(result, 1e-5);
+}
+
+/**
+ * Test for minimizing the Booth function.
+ */
+BOOST_AUTO_TEST_CASE(BoothFunctionTest)
+{
+  PSO optimizer(60, 0.9, 0.5, 0.3, 200, 1e-3);
+  BoothFunction f;
+
+  arma::mat iterate;
+  iterate << 1 << 3;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_CLOSE(result, 0, 1e-3);
+}
+
+/**
+ * Test for the McCormick function.
+ */
+BOOST_AUTO_TEST_CASE(McCormickTest)
+{
+  PSO optimizer;
+  McCormickFunction f;
+
+  arma::mat iterate;
+  iterate << -0.54719 << -1.54719;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_CLOSE(result, -1.9133, 1e-2);
+}
+
+/**
+ * Test for the Eggholder function.
+ */
+BOOST_AUTO_TEST_CASE(EggholderFunctionTest)
+{
+  PSO optimizer;
+  EggholderFunction f;
+
+  arma::mat iterate;
+  iterate << 512 << 404.2319;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_CLOSE(result, -959.6407, 1e-5);
+}
+
+/**
+ * Test for the Easom function.
+ */
+BOOST_AUTO_TEST_CASE(EasomFunctionTest)
+{
+  PSO optimizer;
+  EasomFunction f;
+
+  arma::mat iterate;
+  iterate << 3.14 << 3.14;
+
+  double result = optimizer.Optimize(f, iterate);
+
+  BOOST_REQUIRE_CLOSE(result, -1, 1e-3);
```

Comment:
I think this and some other functions are not being tested properly. In the `EasomFunction` class definition, it is stated: This should optimize to f(x) = -1, at x = [3.14, 3.14].

So, by providing the iterate `[3.14, 3.14]` as input, it would always provide -1 as the output. I tried passing in the iterate as `f.GetInitialPoint()`, but that returns the result as 0.
Maybe this should be dealt with differently. I tried to replicate the test from [adam_test.cpp](https://github.com/mlpack/mlpack/blob/119c410c56a1556749e2d98fc63edf21f40fd942/src/mlpack/tests/adam_test.cpp#L108), but that does not converge to the expected values.

Reply:
The defaults for the PSO are not necessarily good for every given problem, so some values used be tailored to the task at hand. In the context of this PR I think it's fine to select some functions (logistic regression, Bukin function, etc.) and see if we get stable results for each.